packages/core/core/src/RequestTracker.js

```diff
@@ -0,0 +1,427 @@
+    invariant(node.type === 'request');
+    return node.value.result;
+  }
+
+  completeRequest(id: string) {
+    this.graph.invalidNodeIds.delete(id);
+    this.graph.incompleteNodeIds.delete(id);
+  }
+
+  respondToFSEvents(events: Array<Event>): boolean {
+    return this.graph.respondToFSEvents(events);
+  }
+
+  hasInvalidRequests() {
+    return this.graph.invalidNodeIds.size > 0;
+  }
+
+  getInvalidRequests(): Array<Request> {
+    let invalidRequests = [];
+    for (let id of this.graph.invalidNodeIds) {
+      let node = nullthrows(this.graph.getNode(id));
+      invariant(node.type === 'request');
+      invalidRequests.push(node.value);
+    }
+    return invalidRequests;
+  }
+
+  replaceSubrequests(
+    requestId: string,
+    subrequestNodes: Array<RequestGraphNode>
+  ) {
+    this.graph.replaceSubrequests(requestId, subrequestNodes);
+  }
+}
+
+type RequestRunnerOpts = {
+  tracker: RequestTracker,
+  ...
+};
+
+export type RunRequestOpts = {|
+  signal?: ?AbortSignal,
+  parentId?: string
+|};
+
+export type RequestRunnerAPI = {|
+  invalidateOnFileCreate: Glob => void,
+  invalidateOnFileDelete: FilePath => void,
+  invalidateOnFileUpdate: FilePath => void,
+  invalidateOnStartup: () => void,
+  replaceSubrequests: (Array<RequestGraphNode>) => void
+|};
+
+export function generateRequestId(type: string, request: mixed) {
+  return md5FromObject({type, request});
+}
+
+export class RequestRunner<TRequest, TResult> {
+  type: string;
+  tracker: RequestTracker;
+
+  constructor({tracker}: RequestRunnerOpts) {
+    this.tracker = tracker;
+  }
+
+  async runRequest(
+    requestDesc: TRequest,
+    {signal}: RunRequestOpts = {}
+  ): Promise<TResult | void> {
+    let id = this.generateRequestId(requestDesc);
+    let request = {id, type: this.type, request: requestDesc};
+
+    let api = this.createAPI(id);
+
+    this.tracker.trackRequest(request);
+    let result: TResult =
+      // $FlowFixMe
+      (this.tracker.getRequestResult(id): any) ||
+      (await this.run(requestDesc, api));
+    assertSignalNotAborted(signal);
+    // Request may have been removed by a parent request
+    if (!this.tracker.isTracked(id)) {
+      return;
+    }
+    await this.onComplete(requestDesc, result, api);
+    this.tracker.completeRequest(id);
+
+    return result;
+  }
+
+  // unused vars are used for types
+  // eslint-disable-next-line no-unused-vars
+  run(request: TRequest, api: RequestRunnerAPI) {
+    throw new Error(
+      `RequestRunner for type ${this.type} did not implement a run()`
+    );
+  }
+
+  // eslint-disable-next-line no-unused-vars
+  onComplete(request: TRequest, result: TResult, api: RequestRunnerAPI) {
+    throw new Error(
+      `RequestRunner for type ${this.type} did not implement a onComplete()`
+    );
+  }
+
+  generateRequestId(request: TRequest) {
+    return md5FromObject({type: this.type, request});
+  }
+
+  createAPI(requestId: string): RequestRunnerAPI {
```

Comment:
Is the idea here to hide the tracker from these runner implementations? Right now the runners are subclasses, so they’d technically still have access to the tracker directly.

Reply:
This API was a little more useful in my initial implementation, where the runners didn't have access to the tracker. It was also more useful when there was a map of runners passed into the Tracker, because then the API could have a `runRequest` method that hard codes the request id as the parent id for the sub request. That's not really an option now that you create  the runners with the tracker and just call them directly, but I thought the runner map idea was less composable, so I'm okay on losing out on that behavior. Right now the API's only usefulness is that you don't have to pass in the request id to the invalidation functions. Since we will need to pass the request ID in anyway to use as the parent ID for calling sub requests, I may end up getting rid of it.